pkg/compactor/visit_marker.go

```diff
@@ -0,0 +1,194 @@
+	if err := v.updateVisitMarker(ctx); err != nil {
+		level.Error(v.getLogger()).Log("msg", "unable to upsert visit marker file content", "err", err)
+		return
+	}
+	level.Debug(v.getLogger()).Log("msg", "marked pending")
+}
+
+func (v *VisitMarkerManager) MarkCompleted(ctx context.Context) {
+	v.visitMarker.UpdateStatus(v.ownerIdentifier, Completed)
+	if err := v.updateVisitMarker(ctx); err != nil {
+		level.Error(v.getLogger()).Log("msg", "unable to upsert visit marker file content", "err", err)
+		return
+	}
+	level.Debug(v.getLogger()).Log("msg", "marked completed")
+}
+
+func (v *VisitMarkerManager) MarkFailed(ctx context.Context) {
+	v.visitMarker.UpdateStatus(v.ownerIdentifier, Failed)
+	if err := v.updateVisitMarker(ctx); err != nil {
+		level.Error(v.getLogger()).Log("msg", "unable to upsert visit marker file content", "err", err)
+		return
+	}
+	level.Debug(v.getLogger()).Log("msg", "marked failed")
+}
+
+func (v *VisitMarkerManager) DeleteVisitMarker(ctx context.Context) {
+	if err := v.bkt.Delete(ctx, v.visitMarker.GetVisitMarkerFilePath()); err != nil {
+		level.Error(v.getLogger()).Log("msg", "failed to delete visit marker", "err", err)
+		return
+	}
+	level.Debug(v.getLogger()).Log("msg", "visit marker deleted")
+}
+
+func (v *VisitMarkerManager) ReadVisitMarker(ctx context.Context, visitMarker any) error {
+	visitMarkerFile := v.visitMarker.GetVisitMarkerFilePath()
+	visitMarkerFileReader, err := v.bkt.ReaderWithExpectedErrs(v.bkt.IsObjNotFoundErr).Get(ctx, visitMarkerFile)
+	if err != nil {
+		if v.bkt.IsObjNotFoundErr(err) {
+			return errors.Wrapf(errorVisitMarkerNotFound, "visit marker file: %s", visitMarkerFile)
+		}
+		v.visitMarkerReadFailed.Inc()
+		return errors.Wrapf(err, "get visit marker file: %s", visitMarkerFile)
+	}
+	defer runutil.CloseWithLogOnErr(v.getLogger(), visitMarkerFileReader, "close visit marker reader")
+	b, err := io.ReadAll(visitMarkerFileReader)
+	if err != nil {
+		v.visitMarkerReadFailed.Inc()
+		return errors.Wrapf(err, "read visit marker file: %s", visitMarkerFile)
+	}
+	if err = json.Unmarshal(b, visitMarker); err != nil {
+		v.visitMarkerReadFailed.Inc()
+		return errors.Wrapf(errorUnmarshalVisitMarker, "visit marker file: %s, content: %s, error: %v", visitMarkerFile, string(b), err.Error())
+	}
+	level.Debug(v.getLogger()).Log("msg", "visit marker read from file", "visit_marker_file", visitMarkerFile)
+	return nil
+}
+
+func (v *VisitMarkerManager) updateVisitMarker(ctx context.Context) error {
+	visitMarkerFileContent, err := json.Marshal(v.visitMarker)
+	if err != nil {
+		v.visitMarkerWriteFailed.Inc()
+		return err
+	}
+
+	reader := bytes.NewReader(visitMarkerFileContent)
+	if err := v.bkt.Upload(ctx, v.visitMarker.GetVisitMarkerFilePath(), reader); err != nil {
+		v.visitMarkerWriteFailed.Inc()
+		return err
+	}
+	return nil
+}
+
+func (v *VisitMarkerManager) getLogger() log.Logger {
+	return log.With(v.logger, "visit_marker", v.visitMarker.String())
```

Comment:
Any reason why we don't initialize this logger in the constructor? Otherwise, every call of `getLogger` creates a new logger instance

Reply:
The content of visit marker could change along time (ie. status). If initialize logger in constructor, those changes will not reflect in log. If creating logger instance would be concern, how about repeat `"visit_marker", v.visitMarker.String()` in each log statement?

Reply:
Got it thanks for the explanation. Let's keep it then.